File: src/automations/cron_humanize.py
"""Human-readable labels for common 5-field cron expressions (keep in sync with web/src/lib/cronHumanize.ts)."""

from __future__ import annotations

import re

_DOW_NAMES = (
    "Sunday",
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday",
)

_WEEKDAYS = "__weekdays__"
_WEEKENDS = "__weekends__"


def _pad2(n: int) -> str:
    return str(int(n)).zfill(2)


def _format_24h(minute: int, hour: int) -> str:
    return f"{_pad2(hour)}:{_pad2(minute)}"


def _parse_dow_label(dow: str) -> str | None:
    d = dow.strip()
    if d in ("*", "?"):
        return None
    if d in ("1-5", "MON-FRI"):
        return _WEEKDAYS
    if d in ("0,6", "6,0", "SAT,SUN"):
        return _WEEKENDS
    if re.fullmatch(r"\d", d):
        i = int(d, 10)
        if 0 <= i <= 6:
            return _DOW_NAMES[i]
    return None


def humanize_cron_expression(cron: str) -> str | None:
    raw = " ".join((cron or "").strip().split())
    if not raw or raw == "—":
        return None
    parts = raw.split(" ")
    if len(parts) != 5:
        return None

    min_f, hour, dom, month, dow = parts

    m_step = re.fullmatch(r"\*/(\d+)", min_f)
    if m_step and hour == dom == month == dow == "*":
        n = int(m_step.group(1), 10)
        if 1 <= n <= 59:
            return "Every minute" if n == 1 else f"Every {n} minutes"

    if min_f == "0" and hour == dom == month == dow == "*":
        return "Every hour"

    h_step = re.fullmatch(r"\*/(\d+)", hour)
    if min_f == "0" and h_step and dom == month == dow == "*":
        n = int(h_step.group(1), 10)
        if 1 <= n <= 23:
            return "Every hour" if n == 1 else f"Every {n} hours"

    if re.fullmatch(r"\d{1,2}", min_f) and hour == dom == month == dow == "*":
        m = int(min_f, 10)
        if 0 <= m <= 59:
            return f"Every hour at :{_pad2(m)}"

    if re.fullmatch(r"\d{1,2}", min_f) and re.fullmatch(r"\d{1,2}", hour) and dom == month == dow == "*":
        mi = int(min_f, 10)
        h = int(hour, 10)
        if 0 <= h <= 23 and 0 <= mi <= 59:
            return f"Daily at {_format_24h(mi, h)}"

    if re.fullmatch(r"\d{1,2}", min_f) and re.fullmatch(r"\d{1,2}", hour) and dom == month == "*" and dow != "*":
        mi = int(min_f, 10)
        h = int(hour, 10)
        dow_label = _parse_dow_label(dow)
        if dow_label and 0 <= h <= 23 and 0 <= mi <= 59:
            t = _format_24h(mi, h)
            if dow_label == _WEEKDAYS:
                return f"Weekdays at {t}"
            if dow_label == _WEEKENDS:
                return f"Weekends at {t}"
            return f"Weekly on {dow_label} at {t}"

    if min_f == "0" and hour == "0" and re.fullmatch(r"\d{1,2}", dom) and month == dow == "*":
        day = int(dom, 10)
        if 1 <= day <= 31:
            return f"Monthly on day {day} at 00:00"

    return None

File: web/src/lib/cronHumanize.ts
/**
 * Turn common 5-field cron expressions into short English for automation UI.
 * Unknown shapes return null so callers can show the raw expression.
 */
const DOW_NAMES = [
  "Sunday",
  "Monday",
  "Tuesday",
  "Wednesday",
  "Thursday",
  "Friday",
  "Saturday",
] as const;

function pad2(n: number): string {
  return String(Math.trunc(n)).padStart(2, "0");
}

function format24h(minute: number, hour: number): string {
  return `${pad2(hour)}:${pad2(minute)}`;
}

/** Internal markers for grouped days-of-week. */
const _WEEKDAYS = "__weekdays__";
const _WEEKENDS = "__weekends__";

function parseDowLabel(dow: string): string | null {
  const d = dow.trim();
  if (d === "*" || d === "?") return null;
  if (d === "1-5" || d === "MON-FRI") return _WEEKDAYS;
  if (d === "0,6" || d === "6,0" || d === "SAT,SUN") return _WEEKENDS;
  if (/^\d$/.test(d)) {
    const i = parseInt(d, 10);
    if (i >= 0 && i <= 6) return DOW_NAMES[i] ?? null;
  }
  return null;
}

export function humanizeCronExpression(cron: string): string | null {
  const raw = cron.trim().replace(/\s+/g, " ");
  if (!raw || raw === "—") return null;
  const parts = raw.split(" ");
  if (parts.length !== 5) return null;

  const [min, hour, dom, month, dow] = parts;

  // */N * * * * — every N minutes
  const minStep = /^\*\/(\d+)$/.exec(min);
  if (minStep && hour === "*" && dom === "*" && month === "*" && dow === "*") {
    const n = parseInt(minStep[1], 10);
    if (n >= 1 && n <= 59) {
      if (n === 1) return "Every minute";
      return `Every ${n} minutes`;
    }
  }

  // 0 * * * * — top of each hour
  if (min === "0" && hour === "*" && dom === "*" && month === "*" && dow === "*") {
    return "Every hour";
  }

  // 0 */N * * * — every N hours on the hour
  const hourStep = /^\*\/(\d+)$/.exec(hour);
  if (min === "0" && hourStep && dom === "*" && month === "*" && dow === "*") {
    const n = parseInt(hourStep[1], 10);
    if (n >= 1 && n <= 23) {
      if (n === 1) return "Every hour";
      return `Every ${n} hours`;
    }
  }

  // N * * * * — same minute each hour (not */n)
  if (/^\d{1,2}$/.test(min) && hour === "*" && dom === "*" && month === "*" && dow === "*") {
    const m = parseInt(min, 10);
    if (m >= 0 && m <= 59) {
      return `Every hour at :${pad2(m)}`;
    }
  }

  // M H * * * — daily at fixed time (calendar day)
  if (/^\d{1,2}$/.test(min) && /^\d{1,2}$/.test(hour) && dom === "*" && month === "*" && dow === "*") {
    const mi = parseInt(min, 10);
    const h = parseInt(hour, 10);
    if (h >= 0 && h <= 23 && mi >= 0 && mi <= 59) {
      return `Daily at ${format24h(mi, h)}`;
    }
  }

  // M H * * DOW — weekly on a named day / weekdays
  if (/^\d{1,2}$/.test(min) && /^\d{1,2}$/.test(hour) && dom === "*" && month === "*") {
    const mi = parseInt(min, 10);
    const h = parseInt(hour, 10);
    const dowLabel = parseDowLabel(dow);
    if (dowLabel && h >= 0 && h <= 23 && mi >= 0 && mi <= 59) {
      const t = format24h(mi, h);
      if (dowLabel === _WEEKDAYS) return `Weekdays at ${t}`;
      if (dowLabel === _WEEKENDS) return `Weekends at ${t}`;
      return `Weekly on ${dowLabel} at ${t}`;
    }
  }

  // 0 0 D * * — monthly on day D at midnight
  if (min === "0" && hour === "0" && /^\d{1,2}$/.test(dom) && month === "*" && dow === "*") {
    const day = parseInt(dom, 10);
    if (day >= 1 && day <= 31) {
      return `Monthly on day ${day} at 00:00`;
    }
  }

  return null;
}

/** Re-write legacy "Scheduled run (cron in tz)." lines using {@link humanizeCronExpression}. */
export function prettifyScheduledTriggerSummary(summary: string): string {
  const t = summary.trim();
  const re = /^Scheduled run \((.+?) in (.+?)\)\s*\.?\s*$/;
  const m = t.match(re);
  if (!m) return summary;
  const inner = m[1].trim();
  const tz = m[2].trim();
  const human = humanizeCronExpression(inner);
  return `Scheduled run (${human ?? inner} in ${tz}).`;
}
